src/pages/Admin/MapPreview/MapPreview.tsx

```diff
@@ -1,62 +1,83 @@
-import React, { useRef, useState } from "react";
-
-import { MapPreviewProps } from "./MapPreview.types";
-
-import * as S from "./MapPreview.styles";
-import BackgroundSelect from "../BackgroundSelect";
+import React, { useMemo, useRef, useState } from "react";
+import { isEqual } from "lodash";
+import { DndProvider } from "react-dnd";
+import { HTML5Backend } from "react-dnd-html5-backend";
 
 import { updateRoom, updateVenue_v2 } from "api/admin";
+
 import { useUser } from "hooks/useUser";
-import { DndProvider } from "react-dnd";
-import { HTML5Backend } from "react-dnd-html5-backend";
+
 import {
   Container,
   SubVenueIconMap,
 } from "pages/Account/Venue/VenueMapEdition/Container";
-import { isEqual } from "lodash";
 import Legend from "components/atoms/Legend";
+import * as S from "./MapPreview.styles";
+import BackgroundSelect from "../BackgroundSelect";
+import { MapPreviewProps } from "./MapPreview.types";
 
 const MapPreview: React.FC<MapPreviewProps> = ({
   venueName,
   mapBackground,
   rooms,
   venueId,
+  editing,
+  setEditing,
 }) => {
   const { user } = useUser();
-  const [editing, setEditing] = useState<boolean>(false);
+  const [mapRooms, setMapRooms] = useState(rooms);
 
   const roomRef = useRef<SubVenueIconMap>({});
 
-  if (!user) return <></>;
+  const iconsMap = useMemo(
+    () =>
+      mapRooms!.map((room, index: number) => ({
+        width: room.width_percent,
+        height: room.height_percent,
+        top: room.y_percent,
+        left: room.x_percent,
+        url: room.image_url,
+        roomIndex: index,
+      })),
+    [mapRooms]
+  );
+  console.log(rooms, mapRooms);
 
-  const iconsMap = rooms!.map((room, index: number) => ({
-    width: room.width_percent,
-    height: room.height_percent,
-    top: room.y_percent,
-    left: room.x_percent,
-    url: room.image_url,
-    roomIndex: index,
-  }));
+  if (!user) return <></>;
 
   const handleOnChange = (val: SubVenueIconMap) => {
     if (!isEqual(roomRef.current, val)) {
       roomRef.current = val;
+      const normalizeRooms = Object.values(val).map((room, index) => ({
+        ...rooms![index],
```

Comment:
This looks a bit hacky. is it a WIP implementation or is it going to production?

Reply:
What's the purpose/intent of this bit of code?

Reply:
@0xdevalias The purpose is to create a new array of rooms but with the new updated positions and set them in the state. That way I can use the updated state whilst the locations are being updated otherwise, after the first one is updated the props change and the rooms go back to their old positions. It's a very unpleasant UI flicker.

Reply:
It looks like the rooms (`roomRef`) is an `object`, but the way you are getting the values and mapping through it makes it feel like it should be an `array`?

I don't deeply have my 🧠  in this code to be able to suggest if there is a better way to handle this. Just feels like it's a little convoluted at the moment maybe? Happy to go with your / @mike-lvov 's judgement on it though.